[spv-out] Always give structs with runtime arrays a Block decoration

TheoDulka opened the pull request:
Fixes #2322, and so will fix https://github.com/bevyengine/bevy/issues/8733 and fix https://github.com/webonnx/wonnx/issues/161

I'm sure the code can be polished and expanded on to take `BufferBlock`s into account, but I just based it off of https://vulkan.lunarg.com/doc/view/1.3.246.1/windows/1.3-extensions/vkspec.html#interfaces-resources-descset, table 21. Where uniform buffers, storage buffers, and inline uniform blocks all use the `Block` decoration.

teoxoy reviewed (changes requested):
Looks good but I think we can now remove:

https://github.com/gfx-rs/naga/blob/0491d39232bafc30962f6ea240ab6ed457282a62/src/back/spv/writer.rs#L1657-L1669

Although:

![image](https://github.com/gfx-rs/naga/assets/28601907/76726ce3-765e-440d-8129-1b15a1e36fe4)

Do the validation layers not require that the struct is used by a variable in the storage address space despite it seemingly being required by the VUID?

TheoDulka commented:
Removing the entire block decorating chunk would cause a problem, because this
https://github.com/gfx-rs/naga/blob/0491d39232bafc30962f6ea240ab6ed457282a62/tests/out/spv/binding-buffer-arrays.spvasm#L19 gets removed, leading to
```
error: line 35: Target of NonWritable decoration is invalid: must point to a storage image, uniform block, or storage buffer
  %8 = OpVariable %_ptr_StorageBuffer__arr__struct_6_uint_10 StorageBuffer
```
as an error.

Instead, I removed the extra decorating of the non-`BindingArray`s, which seems to work.

---

For your other comment, from what I can tell (minus `binding-buffer-arrays.wgsl`) the `Block` decorated structs are also in a `StorageBuffer`. An example in boids:

[`boids.spvasm`](https://github.com/gfx-rs/naga/blob/master/tests/out/spv/boids.spvasm#L23-L77)
```
OpName %9 "Particles"
...
OpDecorate %9 Block
...
%17 = OpTypePointer StorageBuffer %9
```
from
https://github.com/gfx-rs/naga/blob/0491d39232bafc30962f6ea240ab6ed457282a62/tests/in/boids.wgsl#L18-L20

teoxoy commented:
> Removing the entire block decorating chunk would cause a problem, because this
> 
> https://github.com/gfx-rs/naga/blob/0491d39232bafc30962f6ea240ab6ed457282a62/tests/out/spv/binding-buffer-arrays.spvasm#L19
> 
> gets removed, leading to
> ```
> error: line 35: Target of NonWritable decoration is invalid: must point to a storage image, uniform block, or storage buffer
>   %8 = OpVariable %_ptr_StorageBuffer__arr__struct_6_uint_10 StorageBuffer
> ```
> 
> as an error.
> 
> Instead, I removed the extra decorating of the non-`BindingArray`s, which seems to work.

I see, that makes sense; the comment was out of date. Can you update it or remove it?

> For your other comment, from what I can tell (minus `binding-buffer-arrays.wgsl`) the `Block` decorated structs are also in a `StorageBuffer`. An example in boids:
> 
> [`boids.spvasm`](https://github.com/gfx-rs/naga/blob/master/tests/out/spv/boids.spvasm#L23-L77)
> 
> ```
> OpName %9 "Particles"
> ...
> OpDecorate %9 Block
> ...
> %17 = OpTypePointer StorageBuffer %9
> ```
> 
> from
> 
> https://github.com/gfx-rs/naga/blob/0491d39232bafc30962f6ea240ab6ed457282a62/tests/in/boids.wgsl#L18-L20

Right but what I meant is if the struct doesn't get used at all; is it fine to always decorate them with `Block` even if unused?

In your example in https://github.com/gfx-rs/naga/issues/2322, the VUID gets triggered because we are missing the `Block` decoration but it might still get triggered even with the `Block` decoration since it's unused by a global variable in the `StorageBuffer` address space; at least that's how I'm interpreting the VUID.

teoxoy commented:
Wait, I don't understand why the VUID gets triggered, the validation code for this VUID is inside `ValidateVariable`

https://github.com/KhronosGroup/SPIRV-Tools/blob/9b923f7cc3dde6e1a4886b577677e52c3093ffcc/source/val/validate_memory.cpp#L758-L762

It only checks global variables. And we already decorate structs that have a runtime array if they are used by a global variable.

Can you share a WGSL shader that when ran through naga's SPV backend fails validation with the VUID? Or even better, add it as a snapshot test to the PR.

teoxoy commented:
My main question is why does the current code in master trigger the validation error at all?

The current code already writes the `Block` decoration on a struct containing the runtime array if used by a global variable.

Maybe I'm missing something and wanted to reproduce it locally.

TheoDulka commented:
Ahh I misunderstood, I see what you mean now. Yeah it does seem to only trigger when it's not used by a global variable. The bevy issue is from this bevy file [mesh_view_types.wgsl](https://github.com/bevyengine/bevy/blob/02b520b4e85fa95f1f4a85656a677e3fc2155f36/crates/bevy_pbr/src/render/mesh_view_types.wgsl#L91-L93), where the file doesn't use `PointLights` in a global variable, as it's used for importing the types. I assume it's the same case with the wonnx issue, `ArrayVector` might not be used. So I guess the validation checks if the struct has a `Block` decoration regardless of use in variables? The reason is beyond me.

I didn't realize my code just ended up moving the `Block` decorations, as the runtime arrays are already accounted for, but only when they're used in a global variable. When they aren't, that's when the validation error shows up. At this point my code handles runtime arrays in structs regardless of if the struct is used.

---

I added [runtime-array-in-unused-struct.wgsl](https://github.com/gfx-rs/naga/blob/adbb8063617c15b41dd585def84852c132920717/tests/in/runtime-array-in-unused-struct.wgsl) to the snapshots and changed the comment of the storage binding array decorating.

TheoDulka commented:
@teoxoy I found out why it doesn't matter if it's in a variable, because the error from the issues is spit out when checking the types, in `ValidateTypeStruct`

https://github.com/KhronosGroup/SPIRV-Tools/blob/9b923f7cc3dde6e1a4886b577677e52c3093ffcc/source/val/validate_type.cpp#L353-L360

The link you sent was for a different validation error when checking variables

> Wait, I don't understand why the VUID gets triggered, the validation code for this VUID is inside `ValidateVariable`
> 
> https://github.com/KhronosGroup/SPIRV-Tools/blob/9b923f7cc3dde6e1a4886b577677e52c3093ffcc/source/val/validate_memory.cpp#L758-L762

teoxoy commented:
Ah, I see! Thank you for tracking it down!

I think we can go ahead with this PR and omit unused structs in the future if it becomes an issue that we are decorating them despite them being unused.

Related:
https://github.com/KhronosGroup/SPIRV-Tools/pull/5094
https://github.com/KhronosGroup/glslang/issues/2439
https://github.com/KhronosGroup/glslang/pull/3130
